### PR TITLE
feat: moved server code to run in worker

### PR DESCRIPTION
**Pull request by nadav-ab**

### CODE REVIEW

**Comment by PeterShershov** [comment]
reviewed by me and @idoros. @AviVahl please take another look

File: packages/app-core/src/define-app-types.ts [hunk 1]
```diff
@@ -209,6 +227,10 @@ export interface IPrepareAppOptions {
     onManifestUpdate: (appProps: IAppManifest<any>) => void;
     fsApi: FSApi;
 }
+export interface ICallServerMethodOptions {
```
Comment (AviVahl) [inline packages/app-core/src/define-app-types.ts:230]: calling these Options is confusing

File: packages/define-remix-app/src/define-remix-app.tsx [hunk 2]
```diff
@@ -180,6 +184,7 @@ export default function defineRemixApp({ appPath, routingPattern }: IDefineRemix
                 extraData: {
                     parentLayouts,
                     routeId: filePathToRouteId(appDir, pageModule),
+                    exportNames: varNames.size ? ['loader', 'meta', 'default'] : ['meta', 'default'],
```
Comment (AviVahl) [inline packages/define-remix-app/src/define-remix-app.tsx:187]: unclear. only if a page has variables it uses a loader?

File: packages/define-remix-app/src/define-remix-app.tsx [hunk 3]
```diff
@@ -211,6 +223,45 @@ export default function defineRemixApp({ appPath, routingPattern }: IDefineRemix
...
+        callServerMethod: async ({ importModule }, filePath, methodName, args) => {
+            if (methodName === 'loader' || methodName === 'action') {
+                args = [
+                    {
+                        params: (args as [DeserializedLoaderArgs])[0].params,
+                        request: deserializeRequest((args as [DeserializedLoaderArgs])[0].request),
+                    },
+                ];
+            }
+            const loader = importModule(filePath);
```
Comment (AviVahl) [inline packages/define-remix-app/src/define-remix-app.tsx:235]: it is unclear why a function that calls the server imports a module locally.

File: packages/define-remix-app/src/define-remix-app.tsx (hunk 2 above)
Comment (idoros) [reply packages/define-remix-app/src/define-remix-app.tsx:187]: seems that way if you look at the template: https://github.com/wixplosives/codux-core/blob/10ac7996091b2ae507ec09a4f01505739cbaecc8/packages/define-remix-app/src/page-template.ts#L3

File: packages/define-remix-app/src/define-remix-app.tsx [hunk 4]
```diff
@@ -211,6 +223,45 @@ export default function defineRemixApp({ appPath, routingPattern }: IDefineRemix
                 />
             );
         },
+        callServerMethod: async ({ importModule }, filePath, methodName, args) => {
+            if (methodName === 'loader' || methodName === 'action') {
+                args = [
```
Comment (AviVahl) [inline packages/define-remix-app/src/define-remix-app.tsx:228]: unclear why the args are being overridden and deserialized at this point.
also, args is unknown[], so adjusting these values here seems overly specific and unsafe

File: packages/define-remix-app/src/define-remix-app.tsx [hunk 5]
```diff
@@ -211,6 +223,45 @@ export default function defineRemixApp({ appPath, routingPattern }: IDefineRemix
...
+                    {
+                        params: (args as [DeserializedLoaderArgs])[0].params,
+                        request: deserializeRequest((args as [DeserializedLoaderArgs])[0].request),
+                    },
+                ];
+            }
+            const loader = importModule(filePath);
+            const moduleRequest = (await loader.moduleResults) as IResults<{
+                [key: string]: (...args: unknown[]) => unknown;
```
Comment (AviVahl) [inline packages/define-remix-app/src/define-remix-app.tsx:237]: this cast in unsafe. should be checked in runtime to see value is actually a function on an object.

**Review by AviVahl — COMMENTED** [review]
writing code this way, where there are no actual tests checking usage next to it, makes evaluating the correctness of this logic virtually impossible.

most of the new logic branches are tested.

in addition, the changes relating the deferred value should have probably gone to their own PR.

**Comment by AviVahl** [comment]
more confusing is the fact that `callServerMethod` **might** be the part that runs in the worker (I'm not sure TBH)... and next to it is `<App />` that runs in the preview?

mixing code this way makes it super hard to follow/understand/read, and causes over-fetching in the client.

File: packages/define-remix-app/src/define-remix-app.tsx [hunk 6]
```diff
@@ -23,12 +26,13 @@ import {
...
     routePathId,
     RoutingPattern,
+    serizalizeResponse,
     toCamelCase,
 } from './remix-app-utils';
 import { manifestToRouter } from './manifest-to-router';
 import { parentLayoutWarning } from './content';
 import { pageTemplate } from './page-template';
-
+import { isDeferredData } from '@remix-run/router';
```
[the remark this answers is not in the document]
Comment (PeterShershov) [reply packages/define-remix-app/src/define-remix-app.tsx:35]: Added

File: packages/define-remix-app/src/define-remix-app.tsx (hunk 5 above)
Comment (PeterShershov) [reply packages/define-remix-app/src/define-remix-app.tsx]: Added a validation for this and changed the casting.

**Comment by PeterShershov** [comment]
After speaking with @nadav-ab about the comments above:

We expect our users to serialize their requests as messages are passed through communication between environments. The `args` are getting overridden to accommodate that.

> more confusing is the fact that callServerMethod might be the part that runs in the worker (I'm not sure TBH)... and next to it is <App /> that runs in the preview?

It is a common usage to write the client and the server code under the same file especially for meta-frameworks. Currently, we will provide this functionality to the users this way.

> writing code this way, where there are no actual tests checking usage next to it, makes evaluating the correctness of this logic virtually impossible.

Unit tests (client and node) will be added in the next PRs.

File: packages/define-remix-app/src/define-remix-app.tsx (hunk 4 above)
Comment (barak007) [reply packages/define-remix-app/src/define-remix-app.tsx:228]: @PeterShershov do we know what's going on here?
